### DISCUSSION TITLE
Merge with grafana.grafana collection

### DISCUSSION THREAD

**Discussion by rrey** [ansible-collections/community.grafana#278] [Ideas]
Grafana labs recently created a [collection](https://galaxy.ansible.com/grafana/grafana) that leaves inside the [Grafana GitHub Organization](https://github.com/grafana/grafana-ansible-collection).

Having 2 collections for Grafana (grafana.grafana and community.grafana) could be confusing for users and we started a discussion with the grafana.grafana maintainer (@ishanjainn) and started discussing a possible merge of the collections under the grafana roof (The Grafana GitHub org).

Mighty @gundalow suggested that we made the discussion public for transparency and to gather feedback from the community.
So here comes this GitHub Discussion, hoping to benefit from the threaded answer in case a subject start going wild.

If you are a user of Grafana collection (the one or the other) and have any comment regarding the merge don't hesitate to provide feedback through this discussion.

The benefit I see from moving to Grafana Org are:
- Increased visibility 
- Having people at Grafana Labs more involved in the collection maintenance, allowing the collection to stay up to date and working with new Grafana versions.

Let's aim for a super smooth migration where community collection users could simply change the collection namespace from `community` to `grafana` and have everything working.

**Comment by rrey** [comment]
# TODO list to prepare the merge

First observations from @ishanjainn:
> 1. Grafana dashboards, data sources and folders modules are present in both the collections. These three modules in grafana.grafana collection don't work for OSS grafana and I believe these modules in community.grafana do work for Grafana Cloud (Pls correct me If I am wrong) so we should use ones in community.grafana.
> 2. Similar for Tests and CI workflows, We will have to merge them such that they work for both collections.

So I guess one of the first things to do would be to:
- Identify modules that are supposed to work only with grafana cloud (because feature is only available there) and the one that should work everywhere.
- choose a naming convention for modules that works only with grafana cloud to make it super easy to identify them. `grafana_cloud_xxxx` seems good enough imo
- For modules that should work everywhere, choose between community and grafana one and add integration tests to cover cloud and oss deployments. 

Note for maintainers : reply to this thread with things you have in mind so that we know where to find things that should result in a GitHub issue.

**Reply by ishanjainn** [reply]
Regarding naming conventions, It will be nice to keep the modules that work across both Cloud and OSS, For example Dashboards as grafana.dashboard instead of grafana.grafana_dashboard to avoid repetition and ultimately have shorter names. For cloud, Yes having a `cloud_` prefix as it is in grafana.grafana should be good to differentiate.

**Comment by rrey** [comment]
# Migration path for collection users

We will have to announce depreciation of the abandoned collection the best ways possible, I see at least:
 - Bullhorn newsletter
 - Depreciation message in documentation
 - Depreciation message at runtime (Check with @gundalow if there is recommendation around this)

The depreciation message should be displayed for a long period to allow people to move to the new collection without being forced to.
The message will have to explicitly reference the new collection name/location.

**Comment by ishanjainn** [comment]
### Thoughts on how we can proceed
src - [community.grafana repo](https://github.com/ansible-collections/community.grafana)
dest - [grafana.grafana repo](https://github.com/grafana/grafana-ansible-collection)

1. We create a branch (For example merge) and do all merge over the branch (Create sub-branches where necessary)
2. We start with `plugin` folder
  i. We can straight away copy all folders from `src` to `dest` except the **modules** folder as `dest` only has modules folder 
  ii. It will be great if we can make sure the dashboard, datasource and folder modules in community.grafana work for Grafana Cloud aswell before we the next step
  iii. Preferably copy over the contents from [grafana_dashboard.py](https://github.com/ansible-collections/community.grafana/blob/main/plugins/modules/grafana_dashboard.py), [grafana_datasource.py](https://github.com/ansible-collections/community.grafana/blob/main/plugins/modules/grafana_datasource.py), [grafana_folder.py](https://github.com/ansible-collections/community.grafana/blob/main/plugins/modules/grafana_folder.py) to [dashboard.py](https://github.com/grafana/grafana-ansible-collection/blob/main/plugins/modules/dashboard.py), 
     [datasource.py](https://github.com/grafana/grafana-ansible-collection/blob/main/plugins/modules/datasource.py), [folder.py ](https://github.com/grafana/grafana-ansible-collection/blob/main/plugins/modules/dashboard.py) respectively. It would be better to keep module names as grafana.dashboard as its the same for both OSS and Cloud. And for Cloud modules, We can keep grafana.cloud_plugin (As its defined right now)
  iv. We should copy the rest of the files from `src` to `dest` as they aren't present in `dest`

3. For the tests folder, We should be able to copy over all folders from `src` except  **integrations** to `dest`. 
   i. With integrations/targets/(*modules)/tasks, It will be nice to have a single `main.yml` (as it is in community.grafana) and include other playbooks. This way it will be easier to seperate tests for Grafana Cloud and Grafana OSS.
  
4. For GitHub Actions, Once we have tests folder and plugins folder ready for merge, We could decide on how the actions should run as right now I am not sure how the structure would look at the end. Let me know if you have any ideas

5. Rest of the folders and README can be updated once we are done with moving over the plugins, tests and .github folders

We could create issues on the grafana.grafana repo for the merging over and pick up issues individually as we go along

**Reply by rrey** [reply]
Can you elaborate on 3. ? Some modules have a lot of cases to test (datasource for example) and having a file for each case makes it easier to read.
Testing with Grafana cloud * should * be as simple as changing the Grafana URL, so maybe we can have an integration test target that simply calls another with different params. (to be tested)

**Reply by felixfontein** [reply]
I would say that which tests should be executed in which environment should depend on the `aliases` files in the test directories. That's how such distinctions are usually done with ansible-test. (See for example `shippable/cloud/group1.` vs `shippable/posix/group1` in the community.crypto integration tests. The concrete names are left-overs from when everything got migrated from ansible/ansible, I think they can easily be changed completely.)
(Edit: I've tried changing the group identifiers for community.crypto here, it seems to work just fine, proving that you can basically select whatever you want: https://github.com/ansible-collections/community.crypto/pull/517)

**Reply by ishanjainn** [reply]
> Testing with Grafana cloud * should * be as simple as changing the Grafana URL, so maybe we can have an integration test target that simply calls another with different params. (to be tested)

Yup thats something we can do, I initially mentioned two different files as in OSS tests in community.grafana, I saw username/password being used but for cloud we could only use the api keys. If we are to make it similar, We would need to change for OSS to use api key instead which is not much in my opinion.

**Reply by ishanjainn** [reply]
> I would say that which tests should be executed in which environment should depend on the aliases files in the test directories.

Havent really had first hand experience using it, but Im game if we opt to do it that way too

**Reply by ishanjainn** [reply]
Hey @rrey, To get started with merging, shall we start with just importing all plugins first and see how everything shapes up and then we can merge over the tests, Does that sound good?

**Reply by rrey** [reply]
I would go one plug-in at the time with the tests.
Once it is added we are sure it works properly with OSS and cloud.

**Reply by rrey** [reply]
Also, it would be nice to keep the Git history to have people keep credit for their work

**Reply by ishanjainn** [reply]
Ah, Yes that makes sense and preferable

**Reply by ishanjainn** [reply]
For moving with git history, any particular way you want to go ahead with?

**Reply by ishanjainn** [reply]
Hey @rrey, Once we have this figured, can we create issues in the grafana.grafana repo and pick them up parallelly?

**Reply by rrey** [reply]
@ishanjainn I'll create a PR to migrate datasource module on your side.
Let's work together on this, I have not much time to do it alone unfortunately.

**Reply by rrey** [reply]
I have second thoughts about the merging strategy. I think you were right, we should merge community in your repo and then make required changes modules per modules.
@ishanjainn could you create a branch that we will use to merge everything to avoid impacting your `main` ?

**Comment by rrey** [comment]
@webknjaz
@Lusitaniae
@kongslund
@ggiinnoo
@emcence
@gotmax23
@CWollinger

The `community.grafana` collection is merging with `grafana.grafana` collection. I initiated the work [here](https://github.com/grafana/grafana-ansible-collection/pull/28) but Grafana Labs requires that contributors signs the CLA.
Would you accept to sign the CLA ?

**Reply by gotmax23** [reply]
Admittedly, my contribution was trivial, but I am very much unenthused about this. I don't think existing contributors should be asked to reassign their copyrights, and I don't think a community collection should be replaced with a corporate collection that requires signing copyrights to said corporation. This takes away the community ownership in multiple ways:

1. This collection will be moved out of the ansible-collections org.
2. Each contributor or their employer owns a part of the project through copyright. When one entity owns the copyrights for the entire project, they are free to re-license the project, make parts of it proprietary, or do other things that may not be in the best interest of the community.
3. Contributors who are unable to sign the CLA for whatever reason are locked out from contributing to the project.

**Reply by rrey** [reply]
Thanks for the feedback. I think I was a big fast and did not measure the impact of Grafana CLA.

@gundalow could you advise on this matter. I don't feel confortable with licensing issues.

Both projects are GPL 3, but what do you think of the CLA ?

**Reply by felixfontein** [reply]
CC @Spredzy @Andersson007 since Gundalow isn't in the community team anymore.

**Reply by gotmax23** [reply]
FYI @Andersson007 is on PTO for the rest of the year

**Reply by gundalow** [reply]
It's up to the individuals to decide what's best for their collection. 

One of the objectives when I created github.com/ansible-collections was to ensure that even if maintainers leave we still have the ability to update the collection and add new maintainers in the future. Though I do agree with the reasons [rrey listed](https://github.com/ansible-collections/community.grafana/discussions/278#discussion-4426553), it's good for this collection to move to be alongside the rest of the opensource Grafana work

> The benefit I see from moving to Grafana Org are:
> * Increased visibility
> * Having people at Grafana Labs more involved in the collection maintenance, allowing the collection to stay up to date and working with new Grafana versions.

The licensing requirements for a collection to be included in an `ansible` collection are detailed in [collection_requirements.rst](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#licensing), Red Hat doesn't enforce anything either way about CLA, though we recognise that some communities do find CLA useful.

**Reply by Andersson007** [reply]
the suggestion by @ishanjainn to move grafana.grafana to this org is still relevant?

**Reply by ishanjainn** [reply]
From my understanding of the above thread, We can go ahead as is?
@rrey

**Reply by rrey** [reply]
@ishanjainn is it mandatory that we ask all community collection contributors to sign the CLA ?

As a maintainer I am ok to merge with the Grafana collection which will remain open source.

But asking all historical contributors to sign the CLA will be an issue, I know that some of them would probably not answer.

**Reply by ishanjainn** [reply]
Sorry for the delay, I understand the hesitance. Perhaps for the PRs having the merge changes, We can skip (Ignore it basically) CLA but post the merge is complete we make mandatory, Would that be something better?
I'd still have to check this with the team but first want to check in with you

**Reply by rrey** [reply]
If by "ignore it" you mean not having the mention of people's github handle in the PR, yes that would be better.
Can you disable the CLA bot completely for the whole collections merge PRs ?

**Reply by ishanjainn** [reply]
By "ignore it" I meant we can force merge without signing CLA just for the PRs having merge changes. Once again, From your message it looks like its the same thing you are asking for if i'm correct. If yes, Lets start with merge process (skipping and merging without CLA for merge PRs). Meanwhile, Ill double confirm with the team

**Reply by Lusitaniae** [reply]
Signed the CLA, sorry for the delay

**Reply by ishanjainn** [reply]
Hey @rrey, Following up, Can we restart the merge process?

**Reply by ishanjainn** [reply]
Largely with this PR https://github.com/grafana/grafana-ansible-collection/pull/28

**Reply by webknjaz** [reply]
FWIW, I don't like signing the CLA that would have an effect on future contributions. I don't think I'll be contributing into that repository. Is there a way for the CLA to only allow the contributions that are already made? I think I only contributed the GHA workflow updates, nothing functional in the collection itself. Also, that PR does not include my patches anyway. I see that https://github.com/grafana/grafana-ansible-collection/blob/main/.github/workflows/ci-test.yml is somewhat outdated and could benefit from implementing what's in https://github.com/ansible-collections/community.grafana/blob/main/.github/workflows/ansible-test.yml, though.

**Reply by gotmax23** [reply]
I remain against the CLA. I'm not sure if the Community Steering Committee has pull here, but if we do, I'm **-1** to the merger unless the CLA requirement is dropped.

**Reply by webknjaz** [reply]
I wonder if Graphana would consider an opposite move — to the @ansible-collections org w/o the CLA requirement :thinking:
Their licensing is compatible.

**Reply by gotmax23** [reply]
I've opened https://github.com/ansible-community/community-topics/issues/212

**Comment by Andersson007** [comment]
@rrey hi, a couple of merge related questions:
- Is the merge totally done?
- If yes, should the readme of community.grafana be updated to redirect people to grafana.grafana? 
- Should the collection get marked as deprecated on galaxy?
- Should community.grafana be deprecated and removed from the Ansible package? @felixfontein am i correct that it can be done using the process for [renaming](https://github.com/ansible-community/ansible-build-data#renaming-a-collection) collections?
- Finally we could archive the repo after ^ is done?

**Reply by rrey** [reply]
The merge was abandoned because of the CLA requirement on the Grafana side.

I added people on the project last week to help maintain the project. Hopefully it will help making the project more active. Let's give it a last chance before deprecating it.

**Reply by Andersson007** [reply]
@rrey sure, thanks for taking care of it!
